### PR TITLE
Generate runtime API from metadata

### LINKED ISSUE #97: Generate the client from the metadata
Currently, all client implementations are written by hand.  This works, but is very tedious and won’t scale to large runtimes.

A better option is to generate the client.  This is already done by the TypeScript client implementation, so it should be quite doable for the Rust one.

### PR DESCRIPTION
**Pull request by ascjones**
Implements macros which given the SCALE encoded metadata from a substrate runtime, will generate an api similar to [polkadot.js](https://polkadot.js.org/docs/api/start/api.tx)

Closes #97, and probably many more issues...

# todo

- [x] Generated type derives
  - [x] Consider adding `Clone` derive to all generated types.
  - [x] Configure macro to generate additional default derives for generated types.
  - [x] Specify derives to add to specific generated types?
 - [x] Make tests all pass
 - [x] Resolve all todos in code
 - [x] more todos...

### CODE REVIEW

**Comment by lamafab** [comment]
Btw, I started a similar project for generating interfaces based on metadata (version 13, though) -> https://github.com/lamafab/gekko

Example: https://docs.rs/gekko/0.1.3/gekko/runtime/kusama/extrinsics/balances/struct.Transfer.html

I guess I can deprecate it now 🙃

**Comment by ascjones** [comment]
> Btw, I started a similar project for generating interfaces based on metadata (version 13, though) -> https://github.com/lamafab/gekko
> 
> Example: https://docs.rs/gekko/0.1.3/gekko/runtime/kusama/extrinsics/balances/struct.Transfer.html
> 
> I guess I can deprecate it now upside_down_face

Actually that's a great idea to use generic type parameters in place of the actual types, to get around the lack of type information in the `v13` metadata. I never thought of that!

I like the API there, nice and clean - I might steal some ideas :upside_down_face:

**Comment by ascjones** [comment]
> I am a bit confused, since the mentioned PR removes the frame_metadata crate, yet this pr still seems to depend on it?

frame-metadata has been extracted to it's own crate here: https://github.com/paritytech/frame-metadata/

> What is the expected timeline for this to happen?

I'm working on it as fast as I can: now that `0.9.11` is out we might be able to merge this branch in an incomplete state, but that will be next week at the earliest. In the meantime you can use this branch (when it is compiling)

**Comment by ascjones** [comment]
Thanks, is that using the `scale` file you posted previously? I will try it out when I get a minute

File: cli/src/main.rs
Comment (ascjones) [reply cli/src/main.rs:37]: This comment is picked up by the `StructOpt` derive and displayed in the help.

File: cli/src/main.rs
Comment (ascjones) [reply cli/src/main.rs:65]: Done

File: codegen/src/types.rs
Comment (ascjones) [reply codegen/src/types.rs]: Yes instead of generating a type it will insert the user supplied type path.

File: Cargo.toml
Comment (ascjones) [reply Cargo.toml:49]: Yes. When released. And we can't release this until they are, so hope that is coming soon.

File: codegen/src/types.rs
Comment (ascjones) [reply codegen/src/types.rs]: No actually it is the name of the types mod itself, not the root mod of the runtime api. I have renamed this to `types_mod_ident` and added some docs to the fields.

File: Cargo.toml
Comment (sander2) [reply Cargo.toml:49]: > Yes. When released. And we can't release this until they are, so hope that is coming soon.

@ascjones are you saying that this pr can't be merged until v4.0.0 of the substrate crates are released? Why can't we use `polkadot-v0.9.11`? This is what I use in my fork of this PR, and seems to compile just fine. Also, I assume the connected parachain will also need to be updated to v4.0.0 for this to work? Will this be a major breaking change?

**Comment by sander2** [comment]
Against what binary should the tests be run? I'm trying to figure out if the sudo tests succeed on this pr, because they don't in my code..

File: cli/src/main.rs
Comment (jsdw) [inline cli/src/main.rs:59]: Nit: format could be an enum with a `FromStr` impl so that structopt can handle the error message for you if it's wrong?

File: cli/src/main.rs
Comment (jsdw) [inline cli/src/main.rs:124]: Nit: I wonder whether something like

```
match (url, file) => {
   (None, Some(file)) => {/* from file codegen */},
   (Some(url), None) => {/* from url codegen */},
   (Some(url),Some(file)) => {/* error can't have both */}
}
```

File: codegen/src/api/mod.rs
Comment (jsdw) [inline codegen/src/api/mod.rs:135]: I think we can use `into_iter()` and avoid the cloning now? :)

Comment (jsdw) [reply codegen/src/api/mod.rs:135]: (and perhaps we don't need the `.to_string()` below either; leave as str?)

(just nits, really)

File: codegen/src/api/mod.rs
Comment (insipx) [inline codegen/src/api/mod.rs:213]: Should we scrape all these todos and make them into issues/put them into a project? might be a good starting point

**Review by insipx — COMMENTED** [review]
still have the rest of codegen/ to review, will finish later today

**Comment by ascjones** [comment]
> Against what binary should the tests be run? I'm trying to figure out if the sudo tests succeed on this pr, because they don't in my code.

I've been using the latest build of substrate: `https://releases.parity.io/substrate/x86_64-debian:stretch/latest/substrate/substrate`. Although we should probably pin against a known tag.

File: src/metadata.rs
Comment (ascjones) [reply src/metadata.rs:240]: Yes `frame-metadata` would be the natural place to put it, maybe as a `TryFrom` impl.

File: codegen/src/ir.rs
Comment (insipx) [inline codegen/src/ir.rs:58]: Ah, the type substitutions make sense now!

**Review by insipx — APPROVED** [review]
LGTM, looks really clean imo!

File: codegen/src/api/mod.rs
Comment (ascjones) [reply codegen/src/api/mod.rs]: https://github.com/paritytech/subxt/issues/309

File: macro/src/lib.rs
Comment (ascjones) [reply macro/src/lib.rs:16]: Added to https://github.com/paritytech/subxt/issues/309

Comment (ascjones) [reply macro/src/lib.rs:16]: #309

File: codegen/src/api/mod.rs
Comment (ascjones) [reply codegen/src/api/mod.rs:99]: Yes would have to be done via the macro.

Comment (ascjones) [reply codegen/src/api/mod.rs:99]: #309

File: cli/src/main.rs
Comment (ascjones) [reply cli/src/main.rs:59]: #309

File: cli/src/main.rs
Comment (ascjones) [reply cli/src/main.rs:124]: #309

File: codegen/src/api/mod.rs
Comment (ascjones) [reply codegen/src/api/mod.rs:126]: #309

**Review by dvdplm — APPROVED** [review]

File: codegen/src/api/storage.rs
Comment (ascjones) [reply codegen/src/api/storage.rs]: Indeed it is entirely possible to break this. In the end the compiler will check for clashes for us so I don't see it as a high priority to solve at this point.

**Comment by ascjones** [comment]
Please see https://github.com/paritytech/subxt/issues/309 for follow-ups to this PR

File: Cargo.toml
Comment (dvdplm) [reply Cargo.toml:49]: @sander2 we can definitely merge this (indeed I'm typing this after it's merged) but we can't publish to crates.io with git dependencies, this is what Andrew means by "releasing" here. :)

Comment (seunlanlege) [reply Cargo.toml:49]: Actually, most people will be using polkadot-v0.9.12 for their parachain. I suggest we do the same for subxt. Or at least maintain a different branch that follows the latest polkadot releases.

**Pull request merged**
